### ISSUE TITLE
[PRD] Unified streaming, window-parametrized dense-regions extraction

### ISSUE THREAD

**Issue by clemsgrs** [clemsgrs/slide2vec#192]
## Problem Statement

A researcher running dense feature-grid extraction over whole-slide images — a frozen
tile encoder applied to many sampled ROIs per slide, each turned into a `(d, grid_h,
grid_w)` token grid — sees the job get SIGKILLed (exit 137) part-way through. Host RSS
climbs monotonically across slides (observed: 23 → 34 → 46 → 56 → 66 GB) until it hits the
cgroup ceiling (~80 GB) and the kernel kills the process. The cache persists incrementally
so a relaunch resumes, but it then OOMs again on the next large slide, so the run cannot
complete unattended.

The cause is not a true leak but **unbounded transient host memory that scales with a
slide's ROI count**. The extraction path builds a Python list holding *every* ROI's grid
for a slide and then `np.concatenate`s the whole list into a single array — a second
full-size allocation. ROI counts per slide are heavily skewed (median ~56, max ~4535), and
one Virchow2 grid is ~7 MB, so the worst slide holds ~32 GB in the list plus ~32 GB again
in the concatenate (~63 GB transient for a single slide). glibc retains freed arenas at the
high-water mark, so RSS ratchets upward across successive large slides rather than
returning to baseline.

Separately, this same accumulate-then-return defect exists in **two places**. slide2vec
owns the whole-region dense path; the downstream library (soma) forked that path to add a
sliding-window variant (slide the encoder's native field of view over a larger supervision
tile and blend the sub-grids with a raised-cosine importance map). The fork re-imports
slide2vec scaffolding (padding, autocast context, grid geometry) just to rebuild the same
read → encode loop with one different step, and carries the identical accumulate-concatenate
tail. The OOM therefore exists in both modes; the sliding mode is the one currently
exercised, so that is where it bites. The duplication also means the window/blend math —
which is encoder featurization, not caching — lives at the wrong layer and is depended on
by three separate consumers.

## Solution

slide2vec exposes **one** dense-regions extraction primitive that is parametrized by
window size and **streams** its results: it yields one `(d, grid_h, grid_w)` grid per ROI,
in coordinate order, keeping at most one batch of grids resident at a time. The caller
writes each grid to disk as it arrives, so the whole-slide array never needs to exist and
resident host memory is bounded by `batch_size`, independent of how many ROIs a slide has.

The same primitive serves both modes. With no window size it does a single padded forward
per region (the current whole-region behavior, byte-for-byte). With a window size smaller
than the encoded tile it slides the encoder's native window over the padded region and
blends the sub-grids — the methodological choice that lets a native-field encoder (e.g.
224-px Virchow2/phikon) serve a larger supervision tile without position-embedding
interpolation. The window/blend math moves down into slide2vec so it lives with the encoder
featurization it belongs to, and the downstream fork is deleted: the downstream library
sources ROI coordinates and persists grids (its job), and calls the one slide2vec primitive
to read + encode (slide2vec's job). One streaming function, one place for the math, one
layer seam.

## User Stories

1. As a researcher running dense extraction over a cohort, I want the job to keep host
   memory bounded regardless of how many ROIs a slide has, so that a slide with thousands
   of ROIs does not OOM-kill the run.
2. As a researcher, I want resident memory to stay flat across successive large slides, so
   that a long unattended run completes instead of ratcheting toward the cgroup ceiling.
3. As a researcher, I want a relaunch after any interruption to resume from the persisted
   cache and run to completion, so that I do not have to babysit and restart the job.
4. As a researcher using a native-field encoder (e.g. 224-px) on a larger supervision tile,
   I want sliding-window dense extraction available from the same extraction entry point,
   so that I do not depend on a separately-maintained fork to get it.
5. As a researcher, I want whole-region extraction results to be byte-identical before and
   after this change, so that previously cached grids and downstream supervision are
   unaffected.
6. As a researcher, I want sliding-window extraction results to be byte-identical before and
   after the math moves layers, so that in-flight segmentation/detection work is unaffected.
7. As a researcher, I want both patch-feature grids and CLS-attention grids supported by the
   one streaming primitive, so that either feature kind benefits from the memory fix.
8. As a researcher, I want region reads to stay spacing-aware (read at the requested
   microns-per-pixel and downscale to the exact target size), so that the token grid still
   registers against masks read at the same spacing.
9. As a downstream-library maintainer, I want to call a single slide2vec function with a
   window-size argument instead of branching between a slide2vec call and a local fork, so
   that the extraction code path is one call, not two.
10. As a downstream-library maintainer, I want to delete the forked sliding extractor, so
    that the read+encode loop, padding, and autocast setup are not duplicated across the
    repo boundary.
11. As a downstream-library maintainer, I want the window/blend primitive importable from
    slide2vec, so that every consumer (extraction, pre-cropped extraction, training-time
    re-encode) imports it from one place.
12. As an ML engineer using the training-time re-encode (augmentation) path, I want the
    window/blend primitive to keep working after it moves to slide2vec, so that training is
    not broken by the relocation.
13. As a downstream-library maintainer, I want the dense-grid metadata sidecar to keep
    recording window size and overlap, so that persisted grids remain self-describing.
14. As a downstream-library maintainer, I want the grid geometry (target size, encoded
    size, token-grid shape, padding) to remain owned by slide2vec and read back from the
    persisted sidecar, so that geometry is computed in exactly one place.
15. As a caller, I want the streaming primitive to validate its arguments eagerly (before
    any region is read), so that a bad pad mode or feature kind fails fast at the call site
    rather than deep inside iteration.
16. As a caller, I want each yielded grid to be a standalone contiguous array, so that
    holding one grid does not pin an entire batch's memory alive.
17. As a caller, I want empty coordinates to simply yield nothing, so that I do not have to
    special-case a sentinel empty array.
18. As a caller iterating results, I want grids yielded in coordinate order, so that I can
    zip them against my records and write each to its sample id without reordering.
19. As a slide2vec maintainer, I want the function name to signal that it is a generator,
    so that callers consciously adopt the streaming contract rather than silently relying on
    a returned array.
20. As a slide2vec maintainer, I want the whole-region path to route through the same
    windowed primitive (which short-circuits the no-window case), so that there is no
    separate special-cased encode branch to keep in sync.
21. As a downstream-library maintainer, I want a single slide2vec version that carries the
    new contract and a pin bump on my side, so that the cross-repo migration is explicit and
    documented in release notes.
22. As a documentation reader, I want the models docs corrected where they currently say
    padding and sliding-window policy are entirely the caller's responsibility, so that the
    docs do not contradict slide2vec now owning region-level padding and sliding-window.
23. As a documentation reader, I want the dense-extraction API docs to describe the
    region-level streaming primitive (spacing-aware reads, padding to the patch multiple,
    optional sliding-window blending, the streaming contract), so that the new capability is
    discoverable rather than known only to the downstream library.

## Implementation Decisions

- **One streaming, window-parametrized dense-regions primitive in the slide2vec runtime.**
  The existing whole-region `encode_regions_dense` is replaced by a generator. It is renamed
  to make the generator contract explicit (e.g. `iter_regions_dense`) — a hard rename, no
  back-compat alias kept (consistent with the project's preference for hard replacement over
  deprecated shims). API contract (the decision, encoded as a signature):

  ```
  iter_regions_dense(*, model, device, wsi, coordinates, requested_spacing_um,
                     target_size, tolerance=0.05, pad_mode="reflect",
                     image_pad_value=None, feature_kind="patch_features",
                     attention_blocks=(-1,), attention_include_registers=False,
                     batch_size=1, precision="fp32", dense_transform=None,
                     window_size=None, overlap=0.0) -> Iterator[np.ndarray]
  ```

  Yields one `float32 (d, grid_h, grid_w)` grid per coordinate, in coordinate order.

- **Eager-validate, lazily-iterate.** Argument validation (pad mode, feature kind) and
  geometry computation run at call time and return an inner generator, so invalid arguments
  fail fast at the call site instead of being deferred to the first iteration step.

- **At most one batch resident.** The loop reads a batch, encodes it, moves it to host
  memory, and yields each row as a standalone contiguous copy so the batch array is freed
  once its rows are emitted (a per-row view would otherwise pin the whole batch alive). The
  per-slide list and the trailing `np.concatenate` are removed entirely. Resident host
  memory is therefore bounded by `batch_size`, not by a slide's ROI count.

- **Unify both modes through the windowed primitive.** The window/blend primitive already
  short-circuits the no-window case to a single whole-tile forward that is byte-identical to
  the current whole-region encode. The streaming function always calls the windowed
  primitive with `window_size`, collapsing the previous whole-region special-case into one
  code path.

- **Move the window/blend math down into slide2vec.** The sliding-window encode primitive
  and its geometry resolver (window size/stride/start positions, raised-cosine importance
  map, patch alignment, edge-window handling) move from the downstream library into
  slide2vec's dense runtime, because they are encoder featurization, not caching. The
  stitch math continues to run in fp32. The human-readable "describe the dense mode" helper
  stays a downstream concern but imports the geometry resolver from slide2vec.

- **Grid geometry stays slide2vec-owned.** The resolved layout (target size, encoded size,
  token-grid shape, bottom/right padding) remains computed in slide2vec and read back from
  the persisted sidecar downstream. No schema change to the dense-grid metadata; it
  continues to record window size and overlap.

- **Clean break at every consumer.** All consumers import the window/blend primitive from
  slide2vec: the ROI sliding extractor (absorbed — its fork is deleted and its `run()`
  collapses the whole-region/sliding branch into one streaming call), the pre-cropped dense
  extractor, the training-time re-encode (augmentation) path, and the metadata/describe path
  that uses the geometry resolver.

- **Versioning.** slide2vec takes a minor version bump (new public contract). The downstream
  library bumps its slide2vec pin and updates all import and call sites in the same
  migration. The rename and the relocation are documented in the slide2vec release notes.

- **Documentation.** Correct the models-docs note that currently states padding and
  sliding-window policy are the caller's responsibility: that remains true of the
  normalization-only encoder transform, but slide2vec now provides a region-level streaming
  primitive that handles spacing-aware reads, padding to the patch multiple, and optional
  sliding-window blending. Add or extend the dense-extraction section of the API docs to
  document this region-level primitive (the streaming contract, the `window_size`/`overlap`
  parameters, and the supported feature kinds). The output-layout docs need no change —
  dense grids are still written by the downstream library, not by the slide2vec `Pipeline`.
  The Sphinx docs build must stay green.

## Testing Decisions

A good test here asserts **external behavior** — grid shapes, byte-level parity against a
hand-rolled reference, the streaming contract — never internal structure. The suite runs
fully offline: random-weight tile encoder (no pretrained download) plus an injected fake
WSI reader (no real slide), exactly the established pattern for the existing dense-regions
tests.

- **Primary seam — the streaming primitive (`iter_regions_dense`).** Parametrized over
  `window_size` (no-window vs a window smaller than the encoded tile) and over `feature_kind`
  (patch features vs CLS attention), assert: (1) yields N grids of the expected
  `(d, grid_h, grid_w)` shape for N coordinates, in coordinate order; (2) each yielded grid
  is byte-identical to a direct transform → pad → [window/blend] → encode reference for the
  same region; (3) **streaming/laziness** — a call-counting fake reader shows only one batch
  is read before the first grids are yielded and reads advance one batch at a time (the
  unit-level proof that the per-slide accumulation is gone); (4) bad pad mode / feature kind
  raise at the call site before any read; (5) empty coordinates yield nothing.

- **Secondary seam — the window/blend math primitive and its geometry resolver.** The
  intricate stitch tests relocate from the downstream library into slide2vec's suite
  unchanged: window-start coverage and patch alignment, edge-window appending when the
  stride misses the boundary, the no-window/whole-window short-circuit being byte-identical
  to a single dense encode, raised-cosine overlap blending, and non-overlap block-encoding
  parity on the grid interior. These are kept (not folded away) because their byte-identity
  guarantees are the safety net for moving the math across the layer boundary.

- **Prior art.** The existing dense-regions tests (offline encoder + fake reader, with a
  parity check against a hand-rolled transform+pad+encode) and the downstream sliding-window
  tests are the templates; the new suite extends the former and absorbs the latter.

- **Integration verification (not a unit test).** The flat-RSS proof is the real extraction
  relaunch: resident memory must stay bounded (≈ one batch working set) even while the
  largest-ROI slide is processed, instead of ratcheting toward the cgroup ceiling. This is
  run as the end-to-end check on the motivating workload, separate from the unit suite.

## Out of Scope

- Streaming the **pre-cropped** dense extractor (the one that operates on already-cropped
  tile batches rather than WSI coordinates). After this change it imports the window/blend
  primitive from slide2vec; whether it also adopts streaming is a separate downstream
  decision.
- Any change to the downstream caching/writing layer beyond updating its imports, deleting
  the fork, and collapsing the extraction branch into one streaming call.
- Multi-GPU sharding of dense extraction.
- The 2D dense feature-grid consistency check tracked separately (see Further Notes).
- Removal of any temporary diagnostics in the downstream worktree (downstream housekeeping,
  not part of this primitive).

## Further Notes

- This PRD is the **slide2vec half**. The downstream library needs a companion migration in
  the same window: bump the slide2vec pin, switch every import and call site to the renamed
  streaming primitive and the relocated window/blend primitive, delete the forked sliding
  extractor, and collapse its whole-region/sliding branch into one call.
- An interim downstream patch (turning the forked extractor into a generator) was applied to
  unblock the motivating run on the current architecture. It is **superseded** by this work
  and should be reverted in favor of routing through the slide2vec primitive.
- The motivating incident is the dense-extraction OOM on the segmentation/BEETLE workload
  (tracked downstream, soma issue #127), where RSS ratcheted to ~80 GB and the run was
  SIGKILLed. glibc arena retention is why freed memory does not return to baseline between
  slides, turning a per-slide transient spike into a monotonic climb.
- Related future work: the 2D dense feature-grid consistency check (#190) depends on this
  extraction path being validated, but is not part of this PRD.
- Layer principle reaffirmed: dense extraction (read + encode + window + blend) belongs in
  slide2vec; coordinate sourcing, caching, and writing belong downstream. This change moves
  the one piece (the window/blend math) that was on the wrong side of that line.

**Comment by clemsgrs** [comment]
All three slide2vec slices are merged and shipped in **v5.1.0** (https://github.com/clemsgrs/slide2vec/releases/tag/5.1.0):

- #193 → #196 — streaming `iter_regions_dense` (leak fix + rename)
- #194 → #198 — sliding-window unified in (`runtime/dense_sliding.py`, `window_size`/`overlap`)
- #195 → #199 — docs

The slide2vec half of this PRD is complete. The companion soma migration (pin bump, import from slide2vec, delete the fork) is tracked at clemsgrs/soma#161, now unblocked by the 5.1.0 release.
